In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/titanic/train.csv
/kaggle/input/titanic/test.csv
/kaggle/input/titanic/gender_submission.csv


Lets load the dataset into pandas!

In [2]:
train_df = pd.read_csv("/kaggle/input/titanic/train.csv")
train_df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [3]:
test_df = pd.read_csv("/kaggle/input/titanic/test.csv")
test_df.head()

PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Cabin Embarked  
0  34.5      0      0   330911   7.8292   NaN        Q  
1  47.0      1      0   363272   7.0000   NaN        S  
2  62.0      0      0   240276   9.6875   NaN        Q  
3  27.0      0      0   315154   8.6625   NaN        S  
4  22.0      1      1  3101298  12.2875   NaN        S

# Explaination of classes:

* Passenger ID: to keep traack of them
* Pclass: passenger class, aka socio economic status
* Name: prefaced by title (Mr. = adult male or young male, Mrs. = married woman, Ms = woman with marital status unknown, Master = young male, Miss = young woman or unmarried woman)
* Sex: male or female
* Age: fractional if less than 1 or if estimated
* SibSp: count of siblings/spouse onboard
* Parch: count of parent/child onboard
* Ticket: number/code, seemingly assigned based on which offices they bought it from and in what order. explained in forum [here](https://www.encyclopedia-titanica.org/community/threads/ticket-numbering-system.20348/post-267061)
* Fare: in 1970's british pounds, for the whole ticket which includes all members in group ( SibSp + Parch + 1 )
* Cabin: which room they were assigned. Letter indicates which deck (floor) they were on. A being the upper deck, G was the lowest. A map of the cabins can be found [here](https://www.encyclopedia-titanica.org/titanic-deckplans/)
* Embarked: S = Southampton (England), C = Cherbourg (France), Q = Queenstown (Ireland) (listed in chronological order

In [4]:
test_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  418 non-null    int64  
 1   Pclass       418 non-null    int64  
 2   Name         418 non-null    object 
 3   Sex          418 non-null    object 
 4   Age          332 non-null    float64
 5   SibSp        418 non-null    int64  
 6   Parch        418 non-null    int64  
 7   Ticket       418 non-null    object 
 8   Fare         417 non-null    float64
 9   Cabin        91 non-null     object 
 10  Embarked     418 non-null    object 
dtypes: float64(2), int64(4), object(5)
memory usage: 36.0+ KB


In [5]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [6]:
women = train_df.loc[train_df.Sex == 'female']["Survived"]
rate_women = sum(women)/len(women)

print("% of women who survived:", rate_women)

men = train_df.loc[train_df.Sex == 'male']["Survived"]
rate_men = sum(men)/len(men)

print("% of men who survived:", rate_men)

% of women who survived: 0.7420382165605095
% of men who survived: 0.18890814558058924


In [7]:
train_df['Fare'].isna().sum()

0

In [8]:
# Convert single column to int dtype.
train_df['Fare'] = train_df['Fare'].astype('int')

In [9]:
test_df['Fare'].isna().sum()

1

uh oh! theres a nan value. lets find out who it is

In [10]:
def display_options():
     
    display = pd.options.display
    display.max_columns = 20
    display.max_rows = 500
    display.max_colwidth = 222
    display.width = None
    return None
 
display_options()
display(test_df)

PassengerId  Pclass  \
0            892       3   
1            893       3   
2            894       2   
3            895       3   
4            896       3   
5            897       3   
6            898       3   
7            899       2   
8            900       3   
9            901       3   
10           902       3   
11           903       1   
12           904       1   
13           905       2   
14           906       1   
15           907       2   
16           908       2   
17           909       3   
18           910       3   
19           911       3   
20           912       1   
21           913       3   
22           914       1   
23           915       1   
24           916       1   
25           917       3   
26           918       1   
27           919       3   
28           920       1   
29           921       3   
30           922       2   
31           923       2   
32           924       3   
33           925       3   
34           926       1   
35           927       3   
36           928       3   
37           929       3   
38           930       3   
39           931       3   
40           932       3   
41           933       1   
42           934       3   
43           935       2   
44           936       1   
45           937       3   
46           938       1   
47           939       3   
48           940       1   
49           941       3   
50           942       1   
51           943       2   
52           944       2   
53           945       1   
54           946       2   
55           947       3   
56           948       3   
57           949       3   
58           950       3   
59           951       1   
60           952       3   
61           953       2   
62           954       3   
63           955       3   
64           956       1   
65           957       2   
66           958       3   
67           959       1   
68           960       1   
69           961       1   
70           962       3   
71           963       3   
72           964       3   
73           965       1   
74           966       1   
75           967       1   
76           968       3   
77           969       1   
78           970       2   
79           971       3   
80           972       3   
81           973       1   
82           974       1   
83           975       3   
84           976       2   
85           977       3   
86           978       3   
87           979       3   
88           980       3   
89           981       2   
90           982       3   
91           983       3   
92           984       1   
93           985       3   
94           986       1   
95           987       3   
96           988       1   
97           989       3   
98           990       3   
99           991       3   
100          992       1   
101          993       2   
102          994       3   
103          995       3   
104          996       3   
105          997       3   
106          998       3   
107          999       3   
108         1000       3   
109         1001       2   
110         1002       2   
111         1003       3   
112         1004       1   
113         1005       3   
114         1006       1   
115         1007       3   
116         1008       3   
117         1009       3   
118         1010       1   
119         1011       2   
120         1012       2   
121         1013       3   
122         1014       1   
123         1015       3   
124         1016       3   
125         1017       3   
126         1018       3   
127         1019       3   
128         1020       2   
129         1021       3   
130         1022       3   
131         1023       1   
132         1024       3   
133         1025       3   
134         1026       3   
135         1027       3   
136         1028       3   
137         1029       2   
138         1030       3   
139         1031       3   
140         1032       3   
141         1033       1   
1

It was Mr. Storey. Why? what other info can we use to deduce his ticket price? We can't delete it because its in the testing dataset which we need to match the number of records on kaggle's answers in order to submit and get an accuracy score so lets figure out what to replace it with. Time to do some digging!

Ticket numbers are probably most dependant on where and when they bought it. aka from which office and in what order it was sold. Fare is based on the group on the same ticket number, so for example Master Arthur Rice, 10 years old, has a third class ticket costing about 29, not because his individual ticket cost that much but because he travelled with 4 siblings and 1 mother, accounted for under SibSP and Parch respectively. Profile linked [here](https://www.encyclopedia-titanica.org/titanic-victim/albert-rice.html)

So, in order to use fare as a vaild feature, we must divide by SibSp plus Parch.
The single NaN fare value, we meet a problem because it belongs to Mr. Thomas Storey who did not travel in a group. Finding his profile though, one can see that he travelled with his shipmates [link to profile here](https://www.encyclopedia-titanica.org/titanic-victim/thomas-storey.html). If this were the entire dataset, we could average their fares and use that for Mr. Storey's, however only Carver's record is in this subset, so we copy his (Fare = 7.2500). We could find the larger dataset and average the actual numbers for everyone. I'm not sure what would make best data handling practises here. We could also have just K-nearest neighbour'd based on ticker number or maybe calculated individal fare's and KNN'd that but I think this is okay because even though they are different ages (60's vs 20's) the group were all presumably of similar socioeconomic status (same job, all third class, all forced to board titanic because a strike on their original boat). also i jsut spent a long time digging up info about this guy and am tired.

In [11]:
test_df['Fare'] = test_df['Fare'].fillna(7.25)

In [12]:
test_df['Fare'].isna().sum()

0

In [13]:
test_df.sort_values(by = 'Ticket')

PassengerId  Pclass  \
335         1227       1   
158         1050       1   
236         1128       1   
191         1083       1   
266         1158       1   
372         1264       1   
240         1132       1   
368         1260       1   
402         1294       1   
305         1197       1   
215         1107       1   
355         1247       1   
28           920       1   
403         1295       1   
74           966       1   
218         1110       1   
75           967       1   
407         1299       1   
26           918       1   
182         1074       1   
41           933       1   
131         1023       1   
141         1033       1   
306         1198       1   
323         1215       1   
148         1040       1   
231         1123       1   
67           959       1   
298         1190       1   
44           936       1   
364         1256       1   
150         1042       1   
356         1248       1   
77           969       1   
100          992       1   
177         1069       1   
48           940       1   
357         1249       3   
390         1282       1   
308         1200       1   
118         1010       1   
270         1162       1   
122         1014       1   
34           926       1   
252         1144       1   
272         1164       1   
397         1289       1   
395         1287       1   
50           942       1   
94           986       1   
39           931       3   
242         1134       1   
196         1088       1   
371         1263       1   
146         1038       1   
245         1137       1   
144         1036       1   
404         1296       1   
168         1060       1   
96           988       1   
82           974       1   
411         1303       1   
69           961       1   
53           945       1   
328         1220       2   
128         1020       2   
12           904       1   
287         1179       1   
399         1291       3   
162         1054       2   
353         1245       2   
230         1122       2   
385         1277       2   
190         1082       2   
101          993       2   
178         1070       2   
326         1218       2   
180         1072       2   
16           908       2   
193         1085       2   
275         1167       2   
329         1221       2   
61           953       2   
43           935       2   
164         1056       2   
346         1238       2   
110         1002       2   
301         1193       2   
241         1133       2   
185         1077       2   
84           976       2   
2            894       2   
13           905       2   
213         1105       2   
324         1216       1   
229         1121       2   
387         1279       2   
143         1035       2   
336         1228       2   
322         1214       2   
137         1029       2   
348         1240       2   
109         1001       2   
7            899       2   
78           970       2   
373         1265       2   
238         1130       2   
149         1041       2   
68           960       1   
133         1025       3   
339         1231       3   
104          996       3   
203         1095       2   
283         1175       3   
161         1053       3   
292         1184       3   
333         1225       3   
321         1213       3   
288         1180       3   
136         1028       3   
8            900       3   
311         1203       3   
249         1141       3   
85           977       3   
225         1117       3   
29           921       3   
297         1189       3   
417         1309       3   
171         1063       3   
248         1140       2   
173         1065       3   
337         1229       3   
274         1166       3   
80           972       3   
237         1129       3   
115         1007       3   
116         1008       3   
35           927       3   
332         1224       3   
347         1239       3   
366         1258       3   
17           909       3   
1

In [14]:
# Convert single column to int dtype.
test_df['Fare'] = test_df['Fare'].astype('int')

In [15]:
type(train_df['Fare'][0])

numpy.int64

Now we needd to create a new column Number of People on Ticket, or NPeople by adding sibsp, parch and 1. then divide fare / NPeople, and put that into a new column InFare (individual Fare). This is the fare we will use in our algo.

In [16]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import accuracy_score
from sklearn.model_selection import cross_val_score

In [17]:
y = train_df["Survived"]

features = ["Pclass", "Fare", "Sex", "SibSp", "Parch"]
X = pd.get_dummies(train_df[features])
X_test = pd.get_dummies(test_df[features])

model = RandomForestClassifier()

model.fit(X, y)
predictions = model.predict(X_test)

output = pd.DataFrame({'PassengerId': test_df.PassengerId, 'Survived': predictions})

output.head()


PassengerId  Survived
0          892         0
1          893         0
2          894         0
3          895         0
4          896         1

In [18]:
param_grid = { 
    'n_estimators': [100, 200, 500],
    'max_features': ['auto', 'sqrt', 'log2'],
    'max_depth' : [4,5,6,7,8],
    'criterion' :['gini', 'entropy']
}

cv_model = GridSearchCV(estimator = model, param_grid = param_grid, cv = 5)
cv_model.fit(X, y)

cv_model.best_params_

{'criterion': 'gini',
 'max_depth': 8,
 'max_features': 'log2',
 'n_estimators': 500}

In [19]:
best_model = RandomForestClassifier(
    criterion = 'entropy',
    max_depth = 8,
    max_features = 'sqrt',
    n_estimators = 100,
    random_state = 1)

best_model.fit(X, y)

predictions = best_model.predict(X_test)

output = pd.DataFrame({'PassengerId': test_df.PassengerId, 'Survived': predictions})

output.head()

PassengerId  Survived
0          892         0
1          893         0
2          894         0
3          895         0
4          896         1

In [20]:
output.to_csv('submission.csv', index=False)
print("Your submission was successfully saved!")

Your submission was successfully saved!


In [21]:
print(GridSearchCV(estimator = model, param_grid = param_grid, cv = 5))

GridSearchCV(cv=5, estimator=RandomForestClassifier(),
             param_grid={'criterion': ['gini', 'entropy'],
                         'max_depth': [4, 5, 6, 7, 8],
                         'max_features': ['auto', 'sqrt', 'log2'],
                         'n_estimators': [100, 200, 500]})
